File: January/22- Palindrome Partitioning/22- Palindrome Partitioning (Noura Algohary).cpp
// Author: Noura Algohary
class Solution {
public:
    bool isPalindrome(string s, int st, int end)
    {
        while(st < end)
        {
            if(s[st]!=s[end])
                return false;
            
            st++;
            end--;
        }
        return true;
    }
    void allPossiblePalidrome(vector<vector<string>>& palindrome, string s, int idx, vector<string>& branch)
    {
       // solution is found 
        if(idx == s.size())
        {
            palindrome.push_back(branch); //add solution to other solutions
            return;
        }
            
        for(int i=idx;i<s.size();i++)
        {
            if(isPalindrome(s, idx, i))
            {
                branch.push_back(s.substr(idx, i - idx + 1));  
                allPossiblePalidrome(palindrome, s, i+1, branch);
                branch.pop_back();
            }
        }
    }
    vector<vector<string>> partition(string s) {
        vector<vector<string>> palindrome;
        vector<string> branch;
        allPossiblePalidrome(palindrome, s, 0, branch);
        return palindrome;
    }
};

File: 01- January/22- Palindrome Partitioning/22- Palindrome Partitioning (Noura Algohary).py
# Author:Noura Algohary
class Solution:
    def isValid(self, s, st, end) -> bool:
        while st < end:
            if s[st] != s[end]:
                return False
            st += 1
            end -= 1
        return True

    def allPossiblePalindrome(self, palindrome, s, idx, branch):
        # solution is found 
        if len(s) == idx:
            palindrome.append(branch[:])    # add solution to other solutions
            return
        
        for i in range(idx, len(s)):
            if self.isValid(s, idx, i):
                branch.append(s[idx:i+1])
                self.allPossiblePalindrome(palindrome, s, i+1, branch)
                branch.pop(-1)

    def partition(self, s: str) -> List[List[str]]:
        palindrome = []
        branch = []
        self.allPossiblePalindrome(palindrome, s, 0, branch)
        return palindrome